File: backend/app/routers/anomalies.py
import uuid
import time
from datetime import datetime
from typing import Dict, Any, List, Optional
from fastapi import APIRouter, HTTPException
from ..database import query_all, query_one, execute_commit
from ..websocket.connection_manager import manager

router = APIRouter(prefix="/api/anomalies", tags=["AI Anomaly Detection & Compliance Engine"])

@router.get("")
def get_all_anomalies(entity_type: Optional[str] = None, risk_level: Optional[str] = None):
    """
    AI Anomaly Detection Engine:
    Analyzes actual DB records (GPS speed jumps, proxy attendance, CCTV downtime, inspection flags)
    and returns detected anomalies. If no anomalies exist in database, returns empty list.
    """
    # Evaluate live anomalies from database events dynamically
    db_anomalies = query_all("SELECT * FROM anomalies ORDER BY created_at DESC")
    
    # Evaluate CCTV downtime anomalies
    offline_cctvs = query_all("SELECT c.*, p.name as project_name FROM cctv_cameras c JOIN projects p ON c.project_id = p.id WHERE c.status = 'OFFLINE'")
    for cam in offline_cctvs:
        aid = f"anom-cctv-{cam['id']}"
        if not any(a["id"] == aid for a in db_anomalies):
            now_str = datetime.utcnow().isoformat() + "Z"
            execute_commit(
                """INSERT INTO anomalies (id, entity_type, entity_id, entity_name, event_type, risk_level, reason, evidence_summary, recommended_action, status, created_at)
                   VALUES (?, 'CCTV', ?, ?, 'CCTV_DOWNTIME', 'MEDIUM', ?, 'Camera offline heartbeat missing', 'Dispatch maintenance engineer & initiate surprise inspection', 'DETECTED', ?)""",
                (aid, cam["id"], cam["camera_name"], f"CCTV camera '{cam['camera_name']}' at project '{cam['project_name']}' is offline", now_str)
            )

    # Re-query all anomalies
    query = "SELECT * FROM anomalies"
    params = []
    conditions = []
    if entity_type:
        conditions.append("entity_type = ?")
        params.append(entity_type)
    if risk_level:
        conditions.append("risk_level = ?")
        params.append(risk_level)
    
    if conditions:
        query += " WHERE " + " AND ".join(conditions)
    query += " ORDER BY created_at DESC"

    results = query_all(query, tuple(params))
    return {
        "status": "SUCCESS",
        "total_anomalies": len(results),
        "anomalies": results,
        "indicator": "OK" if results else "INSUFFICIENT_DATA"
    }

@router.get("/compliance")
def get_project_compliance_scores():
    """
    Calculates compliance scores for all registered DoSJE projects
    based on actual inspection history, attendance rates, and CCTV availability.
    """
    projects = query_all("SELECT * FROM projects")
    compliance_list = []

    for proj in projects:
        comp = query_one("SELECT * FROM compliance_records WHERE project_id = ?", (proj["id"],))
        inspections = query_all("SELECT * FROM inspections WHERE project_id = ?", (proj["id"],))
        cctvs = query_all("SELECT * FROM cctv_cameras WHERE project_id = ?", (proj["id"],))
        anomalies_count = len(query_all("SELECT * FROM anomalies WHERE entity_id = ?", (proj["id"],)))

        # Dynamic Compliance Calculation
        total_cctv = len(cctvs)
        online_cctv = len([c for c in cctvs if c["status"] == "ONLINE"])
        cctv_pct = (online_cctv / total_cctv * 100.0) if total_cctv > 0 else 100.0

        completed_insp = len([i for i in inspections if i["status"] == "COMPLETED"])
        failed_insp = len([i for i in inspections if i["status"] == "FAILED"])

        score = max(0.0, min(100.0, 100.0 - (failed_insp * 15.0) - (anomalies_count * 10.0) + (completed_insp * 2.0)))

        record = {
            "project_id": proj["id"],
            "project_name": proj["name"],
            "scheme": proj["scheme"],
            "ngo_institute": proj["ngo_institute"],
            "state": proj["state"],
            "district": proj["district"],
            "compliance_score": round(score, 1),
            "cctv_availability_pct": round(cctv_pct, 1),
            "total_inspections": len(inspections),
            "completed_inspections": completed_insp,
            "anomaly_count": anomalies_count,
            "status": "HIGH_COMPLIANCE" if score >= 85 else "MODERATE_RISK" if score >= 70 else "NON_COMPLIANT"
        }
        compliance_list.append(record)

    return compliance_list

File: src/api/anomalies.ts
import axios from 'axios';

export interface Anomaly {
  id: string;
  entity_type: 'PROJECT' | 'USER' | 'INSPECTION' | 'CCTV';
  entity_id: string;
  entity_name: string;
  event_type: string;
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  reason: string;
  evidence_summary?: string;
  recommended_action?: string;
  status: string;
  created_at: string;
}

export interface ProjectCompliance {
  project_id: string;
  project_name: string;
  scheme: string;
  ngo_institute: string;
  state: string;
  district: string;
  compliance_score: number;
  cctv_availability_pct: number;
  total_inspections: number;
  completed_inspections: number;
  anomaly_count: number;
  status: string;
}
import { API_BASE_URL } from '../config/apiConfig';

export async function fetchAnomalies(): Promise<{ anomalies: Anomaly[]; indicator: string }> {
  try {
    const res = await axios.get(`${API_BASE_URL}/api/anomalies`);
    return {
      anomalies: res.data?.anomalies || [],
      indicator: res.data?.indicator || 'INSUFFICIENT_DATA',
    };
  } catch (err) {
    console.warn('[Anomalies API]: Error fetching anomalies', err);
    return { anomalies: [], indicator: 'INSUFFICIENT_DATA' };
  }
}

export async function fetchCompliance(): Promise<ProjectCompliance[]> {
  try {
    const res = await axios.get(`${API_BASE_URL}/api/anomalies/compliance`);
    return res.data || [];
  } catch (err) {
    console.warn('[Compliance API]: Error fetching compliance scores', err);
    return [];
  }
}
